File: app/schemas/merchant.py
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import Field

from app.schemas.common import SchemaModel


class PageResponse(SchemaModel):
    total: int = Field(description="总数")
    page: int = Field(description="当前页码")
    size: int = Field(description="每页数量")
    items: list[dict[str, Any]] = Field(default_factory=list, description="数据列表")


class MerchantProfileResponse(SchemaModel):
    merchant_id: str | None = Field(default=None, description="运营平台商户 ID")
    client_id: str = Field(description="Ozon Client-Id")
    shop_name: str | None = Field(default=None, description="店铺名称")
    display_name: str | None = Field(default=None, description="展示名称")
    logo_url: str | None = Field(default=None, description="店铺 Logo")
    status: str = Field(default="active", description="商户状态")
    currency_code: str | None = Field(default=None, description="币种")
    default_warehouse_id: int | None = Field(default=None, description="默认仓库 ID")
    contact_name: str | None = Field(default=None, description="联系人")
    contact_phone: str | None = Field(default=None, description="联系电话")
    contact_email: str | None = Field(default=None, description="联系邮箱")
    last_connected_at: datetime | None = Field(default=None, description="最近连接 Ozon 时间")
    last_error: str | None = Field(default=None, description="最近错误")


class MerchantBootstrapRequest(SchemaModel):
    client_id: str = Field(min_length=1, description="Ozon Client-Id")
    api_key: str = Field(min_length=1, description="Ozon Api-Key")


class MerchantBootstrapResponse(SchemaModel):
    profile: MerchantProfileResponse
    initialized_from_ozon: bool = Field(description="是否通过 Ozon 接口初始化了商户资料")
    credential_valid: bool = Field(default=True, description="凭证是否校验通过")


class DashboardMetric(SchemaModel):
    key: str
    label: str
    value: int


class MerchantDashboardResponse(SchemaModel):
    profile: MerchantProfileResponse
    metrics: list[DashboardMetric]
    task_status_counts: dict[str, int] = Field(default_factory=dict)
    product_status_counts: dict[str, int] = Field(default_factory=dict)
    today_event_count: int = 0
    recent_events: list[dict[str, Any]] = Field(default_factory=list)


class MerchantProductItem(SchemaModel):
    id: int
    offer_id: str
    local_sku: str | None = None
    name: str | None = None
    product_id: int | None = None
    sku: int | None = None
    currency_code: str | None = None
    price: Decimal | None = None
    old_price: Decimal | None = None
    warehouse_id: int | None = None
    warehouse_name: str | None = None
    stock: int | None = None
    cover_image_url: str | None = None
    sync_status: str
    sync_status_label: str
    ozon_status: str | None = None
    last_task_id: int | None = None
    last_error: Any = None
    updated_at: datetime | None = None
    created_at: datetime | None = None


class MerchantTaskItem(SchemaModel):
    id: int
    task_id: int
    action_type: str
    status: str
    status_label: str
    workflow_status: str | None = None
    workflow_status_label: str | None = None
    total_count: int = 0
    success_count: int = 0
    failed_count: int = 0
    submitted_at: datetime | None = None
    last_polled_at: datetime | None = None
    finished_at: datetime | None = None


class MerchantTaskDetailResponse(SchemaModel):
    task: dict[str, Any]
    items: list[dict[str, Any]] = Field(default_factory=list)


class MerchantTaskEventItem(SchemaModel):
    id: int
    event_type: str
    status: str
    status_label: str
    offer_id: str | None = None
    product_id: int | None = None
    sku: int | None = None
    ozon_task_id: int | None = None
    request_id: str | None = None
    message: str
    error_message: str | None = None
    payload: Any = None
    created_at: datetime


class MerchantTaskEventsResponse(SchemaModel):
    items: list[MerchantTaskEventItem] = Field(default_factory=list)
    next_before_id: int | None = Field(default=None, description="下一页 before_id")
    today_count: int = Field(default=0, description="今日流水数量")

File: merchant-h5/src/api/merchant.ts
import { http } from './http'

export interface PageResult<T> {
  total: number
  page: number
  size: number
  items: T[]
}

export interface MerchantProfile {
  merchant_id: string | null
  client_id: string
  shop_name: string | null
  display_name: string | null
  logo_url: string | null
  status: string
  currency_code: string | null
  default_warehouse_id: number | null
  contact_name: string | null
  contact_phone: string | null
  contact_email: string | null
  last_connected_at: string | null
  last_error: string | null
}

export interface AppActivationCheckResult {
  bound: boolean
  expired: boolean
  activation_required: boolean
  status: string | null
  reason: string | null
  client_id: string | null
  expires_at: string | null
}

export interface AppActivationBindResult {
  bound: boolean
  client_id: string
  expires_at: string
  profile: MerchantProfile
}

export interface AppActivationCodeCreateResult {
  device_id: string
  client_id: string
  activation_code: string
  expires_at: string
  status: string
}

export interface AppActivationCodeItem {
  id: number
  device_id: string
  mac_address: string | null
  client_id: string
  activation_code: string | null
  status: string
  expires_at: string
  activated_at: string | null
  last_seen_at: string | null
  created_at: string
  updated_at: string
}

export interface AppActivationCodeListResult {
  items: AppActivationCodeItem[]
}

export interface DashboardMetric {
  key: string
  label: string
  value: number
}

export interface TaskEvent {
  id: number
  event_type: string
  status: string
  status_label: string
  offer_id: string | null
  product_id: number | null
  sku: number | null
  ozon_task_id: number | null
  request_id: string | null
  message: string
  error_message: string | null
  payload: unknown
  created_at: string
}

export interface DashboardData {
  profile: MerchantProfile
  metrics: DashboardMetric[]
  task_status_counts: Record<string, number>
  product_status_counts: Record<string, number>
  today_event_count: number
  recent_events: TaskEvent[]
}

export interface ProductItem {
  id: number
  offer_id: string
  local_sku: string | null
  name: string | null
  product_id: number | null
  sku: number | null
  currency_code: string | null
  price: string | number | null
  old_price: string | number | null
  warehouse_id: number | null
  warehouse_name: string | null
  stock: number | null
  cover_image_url: string | null
  sync_status: string
  sync_status_label: string
  ozon_status: string | null
  last_task_id: number | null
  last_error: unknown
  updated_at: string | null
  created_at: string | null
}

export interface TaskEventsResult {
  items: TaskEvent[]
  next_before_id: number | null
  today_count: number
}

export interface TranslateTextItem {
  text: string
  translated_text: string
  source_language: string | null
  target_language: 'zh' | 'ru' | 'en'
  translated: boolean
  error?: string | null
}

export interface TranslateTextResult {
  items: TranslateTextItem[]
}

export function bootstrapMerchant(payload: { client_id: string; api_key: string }) {
  return http.post<unknown, { profile: MerchantProfile; initialized_from_ozon: boolean; credential_valid: boolean }>(
    '/merchant/bootstrap',
    payload,
  )
}

export function checkActivation(payload: { device_id: string; mac_address?: string | null }) {
  return http.post<unknown, AppActivationCheckResult>('/app-activations/check', payload)
}

export function bindActivation(payload: {
  device_id: string
  mac_address?: string | null
  client_id: string
  api_key: string
  activation_code: string
}) {
  return http.post<unknown, AppActivationBindResult>('/app-activations/bind', payload)
}

export function createActivationCode(
  payload: {
    device_id: string
    mac_address?: string | null
    client_id: string
    api_key: string
    activation_code?: string | null
    valid_days?: number
    expires_at?: string | null
  },
  adminToken: string,
) {
  return http.post<unknown, AppActivationCodeCreateResult>('/app-activations/codes', payload, {
    headers: {
      'X-Admin-Token': adminToken,
    },
  })
}

export function listActivationCodes(adminToken: string) {
  return http.get<unknown, AppActivationCodeListResult>('/app-activations/codes', {
    headers: {
      'X-Admin-Token': adminToken,
    },
  })
}

export function getDashboard() {
  return http.get<unknown, DashboardData>('/merchant/dashboard')
}

export function getProfile() {
  return http.get<unknown, MerchantProfile>('/merchant/profile')
}

export function getProducts(params: Record<string, unknown>) {
  return http.get<unknown, PageResult<ProductItem>>('/merchant/products', { params })
}

export function getTaskEvents(params: Record<string, unknown>) {
  return http.get<unknown, TaskEventsResult>('/merchant/task-events', { params })
}

export function translateTexts(payload: {
  texts: string[]
  target_language: 'zh' | 'ru' | 'en'
  source_language?: string
}) {
  return http.post<unknown, TranslateTextResult>('/merchant/translate', payload)
}
